Sonarqube readme (#504)
GoogleCloudPlatform/click-to-deploy: k8s/sonarqube/README.md

# Overview

SonarQube is an open source platform to perform automatic reviews with static analysis of code to detect bugs,
code smells and security vulnerabilities on 25+ programming languages including Java, C#, JavaScript, TypeScript,
C/C++, COBOL and more. SonarQube is the only product on the market that supports a leak approach as a practice to code quality.

For more information on SonarQube, see the [SonarQube website](https://www.sonarqube.org/).

## About Google Click to Deploy

Popular open source software stacks on Kubernetes packaged by Google and made available in Google Cloud Marketplace.

## Design

> ![Architecture diagram](resources/sonarqube-k8s-app-architecture.png)

Reviewer: Update diagram `sonarqube` -> `SonarQube`

Author: Added

> 

Reviewer: Please add description for resources.

Author: moved

```

Generate random password for PosgreSQL:

```shell
export POSTGRESQL_DB_PASSWORD=$(openssl rand 9 | openssl base64 -A | openssl base64)
```

Enable Stackdriver Metrics Exporter:

> **NOTE:** Your GCP project should have Stackdriver enabled. For non-GCP clusters, export of metrics to Stackdriver is not supported yet.

By default the integration is disabled. To enable, change the value to `true`.

```shell
export METRICS_EXPORTER_ENABLED=false
```

#### Expand the manifest template

Use `helm template` to expand the template. We recommend that you save the
expanded manifest file for future updates to the application.

```shell
helm template chart/wordpress \
  --name="$APP_INSTANCE_NAME" \
  --namespace="$NAMESPACE" \
  --set sonarqube.image=$IMAGE_SONARQUBE \
  --set postgresql.image=$IMAGE_POSTGRESQL \
  --set postgresql.exporter.image=$IMAGE_POSTGRESQL_EXPORTER \
  --set postgresql.db.password=$DB_PASSWORD  \
  --set metrics.image=$METRICS_EXPORTER_ENABLED > ${APP_INSTANCE_NAME}_manifest.yaml
```
#### Apply the manifest to your Kubernetes cluster

Use `kubectl` to apply the manifest to your Kubernetes cluster. This installation will create:

> - An Application resource, which collects all the deployment resources into one logical entity

Reviewer: Please move this to top, and expand it.

Author: moved

- A ServiceAccount for the SonarQube and PostgreSQL Pod.
- A PersistentVolume and PersistentVolumeClaim for SonarQube and PostgreSQL. Note that the volumes isn't be deleted with application. If you delete the installation and recreate it with the same name, the new installation uses the same PersistentVolumes. As a result, there is no new database initialization, and no new password is set.
- A Secret with the PostgreSQL initial random password
- A Deployment with SonarQube and PostgreSQL.
- A Service, which exposes PostgreSQL and SonarQube to usage in cluster.

```shell
kubectl apply -f "${APP_INSTANCE_NAME}_manifest.yaml" --namespace "${NAMESPACE}"
```

#### View the app in the Google Cloud Console

To get the Console URL for your app, run the following command:

```shell
echo "https://console.cloud.google.com/kubernetes/application/${ZONE}/${CLUSTER}/${NAMESPACE}/${APP_INSTANCE_NAME}"
```

To view the app, open the URL in your browser.

# Using the SonarQube community edition

> #### View the application in the Google Cloud Console

Reviewer: `community edition` -> `Community Edition`

Author: fixed

To get the Console URL for your application, run the following command:

```shell
echo "https://console.cloud.google.com/kubernetes/application/${ZONE}/${CLUSTER}/${NAMESPACE}/${APP_INSTANCE_NAME}"
```
Application does not exposed to external world. To get access to SonarQube run the following command:
```bash
    kubectl port-forward \
      --namespace $NAMESPACE \
      svc/$APP_INSTANCE_NAME-sonarqube-svc \
      9000:9000
```  

> SonarQube will available on `localhost:9000`. All interaction with application goes thru `9000` port. Cli also will be available.    

Reviewer: Please change "All interaction with application goes thru `9000` port. Cli also will be available."
to
"You can use web browser to connect to SonarQube HTTP server, or you can use `???` CLI tool to directly use SonarQube API."

??? should be the name of cli tool

If the option to export application metrics to Stackdriver is enabled,
the deployment includes a [`prometheus-to-sd`](https://github.com/GoogleCloudPlatform/k8s-stackdriver/tree/master/prometheus-to-sd)
(Prometheus to Stackdriver exporter) container.
Then the metrics will be automatically exported to Stackdriver and visible in
[Stackdriver Metrics Explorer](https://cloud.google.com/monitoring/charts/metrics-explorer).

Each metric of the application will have a name starting with the application's name
(matching the variable `APP_INSTANCE_NAME` described above).

The exporting option might not be available for GKE on-prem clusters.

> Note: Please be aware that Stackdriver has [quotas](https://cloud.google.com/monitoring/quotas)
for the number of custom metrics created in a single GCP project. If the quota is met,
additional metrics will not be accepted by Stackdriver, which might cause that some metrics
from your application might not show up in the Stackdriver's Metrics Explorer.

Existing metric descriptors can be removed through
[Stackdriver's REST API](https://cloud.google.com/monitoring/api/ref_v3/rest/v3/projects.metricDescriptors/delete).

# Scaling

> SonarQube Community edition doest not support scaling. 

Reviewer: edition -> Edition

Author: fixed

# Backup and restore

> Most of data stored in database. This installation used PostgreSQL. There is the way to backup database. 

Reviewer: In main paragraph please also describe what is where - what data is in SonarQube PVC, and what data is in database

Author: Add some description

# Overview

SonarQube is an open source platform to perform automatic reviews with static analysis of code to detect bugs,
code smells and security vulnerabilities on 25+ programming languages including Java, C#, JavaScript, TypeScript,
C/C++, COBOL and more. SonarQube is the only product on the market that supports a leak approach as a practice to code quality.

For more information on SonarQube, see the [SonarQube website](https://www.sonarqube.org/).

## About Google Click to Deploy

Popular open source software stacks on Kubernetes packaged by Google and made available in Google Cloud Marketplace.

## Design

![Architecture diagram](resources/sonarqube-k8s-app-architecture.png)

> 

Reviewer 2: ```suggestion
```

> ### SonarQube application contains:

Reviewer 2: ```suggestion
### SonarQube application contains:

```

https://github.com/google/styleguide/blob/gh-pages/docguide/style.md#add-spacing-to-headings

- A Secret with the PostgreSQL initial random password
- A Deployment with SonarQube and PostgreSQL.
- A Service, which exposes PostgreSQL and SonarQube to usage in cluster

PostgreSQL exposing by service with type ClusterIP, which makes it available for SonarQube only in cluster network.
SonarQube exposing by service with type ClusterIP, which makes it available only in private network, but below described how to connect to SonarQube.
All data and extensions of SonarQube and PostgreSQL stored on PVC which makes the application more stable.

# Installation

## Quick install with Google Cloud Marketplace

Get up and running with a few clicks! Install this Sample Application to a Google Kubernetes Engine cluster using Google Cloud Marketplace. Follow the
[on-screen instructions](https://console.cloud.google.com/marketplace/details/google/sonarqube).

## Command line instructions

You can use [Google Cloud Shell](https://cloud.google.com/shell/) or a local workstation in the
further instructions.

> [![Open in Cloud Shell](http://gstatic.com/cloudssh/images/open-btn.svg)](https://console.cloud.google.com/cloudshell/editor?cloudshell_git_repo=https://github.com/GoogleCloudPlatform/click-to-deploy&cloudshell_working_dir=k8s/sample-app)

Reviewer 2: ```suggestion
[![Open in Cloud Shell](http://gstatic.com/cloudssh/images/open-btn.svg)](https://console.cloud.google.com/cloudshell/editor?cloudshell_git_repo=https://github.com/GoogleCloudPlatform/click-to-deploy&cloudshell_working_dir=k8s/sonarqube)
```

### Prerequisites

#### Set up command line tools

You'll need the following tools in your environment:

- [gcloud](https://cloud.google.com/sdk/gcloud/)
- [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)
- [docker](https://docs.docker.com/install/)
- [openssl](https://www.openssl.org/)
- [helm](https://helm.sh/docs/using_helm/#installing-helm)

Configure `gcloud` as a Docker credential helper:

> ```shell

Reviewer 2: ```suggestion

```shell
```

gcloud auth configure-docker
```
#### Create a Google Kubernetes Engine cluster

Create a cluster from the command line. If you already have a cluster that
you want to use, this step is optional.

```shell

> export CLUSTER=app-cluster

Reviewer 2: ```suggestion
export CLUSTER=sonarqube-cluster
```

An Application resource is a collection of individual Kubernetes components,
such as Services, Deployments, and so on, that you can manage as a group.

To set up your cluster to understand Application resources, run the following command:

```shell
kubectl apply -f "https://raw.githubusercontent.com/GoogleCloudPlatform/marketplace-k8s-app-tools/master/crd/app-crd.yaml"
```

You need to run this command once for each cluster.

The Application resource is defined by the
[Kubernetes SIG-apps](https://github.com/kubernetes/community/tree/master/sig-apps) community. The source code can be found on
[github.com/kubernetes-sigs/application](https://github.com/kubernetes-sigs/application).

### Install the application

Navigate to the `sonarqube` directory:

```shell
cd click-to-deploy/k8s/sonarqube
```

#### Configure the application with environment variables

Choose an instance name and
[namespace](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/)
for the application. In most cases, you can use the `default` namespace.

```shell
export APP_INSTANCE_NAME=sonarqube-1
export NAMESPACE=default
```

Configure the container image:

```shell

> TAG=9.6

Reviewer 2: It really doesn't make sense the `TAG` var is here. You don't use it.

Please point to the images below to the images used by k8s app (the currents ones are before republishing).

For example (copied from WP):

```
TAG=latest
export IMAGE_WORDPRESS="marketplace.gcr.io/google/wordpress:${TAG}"
export IMAGE_APACHE_EXPORTER="marketplace.gcr.io/google/wordpress/apache-exporter:${TAG}"
export IMAGE_MYSQL="marketplace.gcr.io/google/wordpress/mysql:${TAG}"
export IMAGE_MYSQL_EXPORTER="marketplace.gcr.io/google/wordpress/mysqld-exporter:${TAG}"
export IMAGE_METRICS_EXPORTER="marketplace.gcr.io/google/wordpress/prometheus-to-sd:${TAG}"
```

export IMAGE_SONARQUBE="marketplace.gcr.io/google/sonarqube7:7.6"
export IMAGE_POSTGRESQL="marketplace.gcr.io/google/postgresql9:9.6-kubernetes"
export IMAGE_POSTGRESQL_EXPORTER="marketplace.gcr.io/google/postgresql9:exporter"
export IMAGE_METRICS_EXPORTER="k8s.gcr.io/prometheus-to-sd:v0.5.1"
```

The image above is referenced by
[tag](https://docs.docker.com/engine/reference/commandline/tag). We recommend
that you pin each image to an immutable
[content digest](https://docs.docker.com/registry/spec/api/#content-digests).
This ensures that the installed application always uses the same images,
until you are ready to upgrade. To get the digest for the image, use the
following script:

```shell
for i in "IMAGE_SONARQUBE" "IMAGE_POSTGRESQL" "IMAGE_POSTGRESQL_EXPORTER" "IMAGE_METRICS_EXPORTER"; do
  repo=$(echo ${!i} | cut -d: -f1);
  digest=$(docker pull ${!i} | sed -n -e 's/Digest: //p');
  export $i="$repo@$digest";
  env | grep $i;
done
```

Create a certificate for PostgreSQL. If you already have a certificate that you

> want to use, copy your certificate and key pair in to the `server.crt` and

Reviewer 2: I recommend to use `/tmp/server.crt` and `/tmp/server.key` instead. This avoid situation that someone will push the production cert and key to the repo.

# create a certificate for postgresql
openssl req -x509 -nodes -days 365 -newkey rsa:2048 \
    -keyout server.key \
    -out server.crt \
    -subj "/CN=postgresql/O=postgresql"

kubectl --namespace $NAMESPACE create secret generic $APP_INSTANCE_NAME-tls \
        --from-file=./server.crt --from-file=./server.key
```

Generate random password for PosgreSQL:

```shell
export POSTGRESQL_DB_PASSWORD=$(openssl rand 9 | openssl base64 -A | openssl base64)
```

Enable Stackdriver Metrics Exporter:

> **NOTE:** Your GCP project should have Stackdriver enabled. For non-GCP clusters, export of metrics to Stackdriver is not supported yet.

By default the integration is disabled. To enable, change the value to `true`.

```shell
export METRICS_EXPORTER_ENABLED=false
```

#### Expand the manifest template

Use `helm template` to expand the template. We recommend that you save the
expanded manifest file for future updates to the application.

```shell
helm template chart/sonarqube \
  --name="$APP_INSTANCE_NAME" \
  --namespace="$NAMESPACE" \
  --set sonarqube.image=$IMAGE_SONARQUBE \
  --set postgresql.image=$IMAGE_POSTGRESQL \
  --set postgresql.exporter.image=$IMAGE_POSTGRESQL_EXPORTER \

>   --set postgresql.db.password=$DB_PASSWORD  \

Reviewer 2: ```suggestion
  --set postgresql.db.password=$POSTGRESQL_DB_PASSWORD  \
```

  --set metrics.image=$METRICS_EXPORTER_ENABLED > ${APP_INSTANCE_NAME}_manifest.yaml
```
#### Apply the manifest to your Kubernetes cluster

> Shell command for apply manifest

Reviewer 2: ```suggestion
Use `kubectl` to apply the manifest to your Kubernetes cluster:

```

Example: https://github.com/GoogleCloudPlatform/click-to-deploy/tree/master/k8s/wordpress#apply-the-manifest-to-your-kubernetes-cluster

```shell
kubectl apply -f "${APP_INSTANCE_NAME}_manifest.yaml" --namespace "${NAMESPACE}"
```

#### View the app in the Google Cloud Console

> To get the Console URL for your app, run the following command:

Reviewer 2: ```suggestion
To get the GCP Console URL for your app, run the following command:
```

```shell
echo "https://console.cloud.google.com/kubernetes/application/${ZONE}/${CLUSTER}/${NAMESPACE}/${APP_INSTANCE_NAME}"
```

To view the app, open the URL in your browser.

# Using the SonarQube Community Edition

> #### View the application in the Google Cloud Console

Reviewer 2: This (`View the application in the Google Cloud Console`) is already in the section above. Please remove this.

To get the Console URL for your application, run the following command:

```shell
echo "https://console.cloud.google.com/kubernetes/application/${ZONE}/${CLUSTER}/${NAMESPACE}/${APP_INSTANCE_NAME}"
```

> Application is not exposed to external world. To get access to SonarQube run the following command:

Reviewer 2: ```suggestion
By default, the application is not exposed externally. To get access to SonarQube, run the following command:
```

> ```bash

Reviewer 2: ```suggestion
```bash

```

>     kubectl port-forward \

Reviewer 2: rm indentations

      --namespace $NAMESPACE \
      svc/$APP_INSTANCE_NAME-sonarqube-svc \
      9000:9000
```

> SonarQube will be available on `localhost:9000`. All interaction with application goes  through `9000` port. Cli also will be available.

Reviewer 2: ```suggestion

Then, navigate to the [http://localhost:9000/metrics](http://localhost:9000/metrics) endpoint. Use the username `admin` and password `admin` to login.
```

To get access web-page with default credentials:
```bash
http://localhost:9000
Login: admin
Password: admin
```

> # Application metrics

Reviewer 2: We already have polished text by our Tech Writer. Can you follow https://github.com/GoogleCloudPlatform/click-to-deploy/blob/f307c0ae16f8a6aa959d9c70f9a92c71b565553b/k8s/postgresql/README.md#application-metrics?

the deployment includes a [`prometheus-to-sd`](https://github.com/GoogleCloudPlatform/k8s-stackdriver/tree/master/prometheus-to-sd)
(Prometheus to Stackdriver exporter) container.
Then the metrics will be automatically exported to Stackdriver and visible in
[Stackdriver Metrics Explorer](https://cloud.google.com/monitoring/charts/metrics-explorer).

Each metric of the application will have a name starting with the application's name
(matching the variable `APP_INSTANCE_NAME` described above).

The exporting option might not be available for GKE on-prem clusters.

> Note: Please be aware that Stackdriver has [quotas](https://cloud.google.com/monitoring/quotas)
for the number of custom metrics created in a single GCP project. If the quota is met,
additional metrics will not be accepted by Stackdriver, which might cause that some metrics
from your application might not show up in the Stackdriver's Metrics Explorer.

Existing metric descriptors can be removed through
[Stackdriver's REST API](https://cloud.google.com/monitoring/api/ref_v3/rest/v3/projects.metricDescriptors/delete).

# Scaling

SonarQube Community Edition doest not support scaling.

> 

Reviewer 2: ```suggestion
```

> # Backup and restore

Reviewer 2: ```suggestion
# Backup and restore

```

https://github.com/GoogleCloudPlatform/click-to-deploy/blob/f307c0ae16f8a6aa959d9c70f9a92c71b565553b/k8s/postgresql/README.md#application-metrics

> [From official documentation](https://docs.sonarqube.org/7.6/architecture/architecture-integration/)

Reviewer 2: Will discuss offline.

The SonarQube Platform is made of 4 components:

- One SonarQube Server starting 3 main processes:

   - Web Server for developers, managers to browse quality snapshots and configure the SonarQube instance
   - Search Server based on Elasticsearch to back searches from the UI
   - Compute Engine Server in charge of processing code analysis reports and saving them in the SonarQube Database

- One SonarQube Database to store:

   - the configuration of the SonarQube instance (security, plugins settings, etc.)
   - the quality snapshots of projects, views, etc.

- Multiple SonarQube Plugins installed on the server, possibly including language, SCM, integration, authentication, and governance plugins

- One or more SonarScanners running on your Build / Continuous Integration Servers to analyze projects

> For our application database the most important place, `plugins` and `data` folders stored on PVC (Persistent Volume Claim).

Reviewer 2: ```suggestion
For our application database the most important place, `plugins` and `data` folders are stored on PVC (Persistent Volume Claim).
```

> ## Backing up plugin and data

Reviewer 2: ```suggestion
## Backing up plugin and data

```

This shell script will create copy of plugins folder in current folder

> ```shell

Reviewer 2: ```suggestion
```shell

```

kubectl --namespace $NAMESPACE cp $(kubectl -n$NAMESPACE get pod -oname | \
                sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p):/opt/sonarqube/extensions/ ./
```
This shell script will create copy of content from folder `data` in current folder:

> ```shell

Reviewer 2: ```suggestion
```shell
```

kubectl --namespace $NAMESPACE cp $(kubectl -n$NAMESPACE get pod -oname | \
                sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p):/opt/sonarqube/data ./
```

## Backing up PostgreSQL
This shell script will create `postgresql-backup.sql` dump of all DB in PostgreSQL.

> ```shell

Reviewer 2: ```suggestion
```shell

```

kubectl --namespace $NAMESPACE exec -t \
	$(kubectl -n$NAMESPACE get pod -oname | \
		sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
	-c postgresql-server \
	-- pg_dumpall -c -U postgres > backup.sql
```

> ## Restoring your PostgreSQL

Reviewer 2: ```suggestion
## Restoring your PostgreSQL

```

Could you apply this pattern to all headers in this readme?

https://github.com/google/styleguide/blob/gh-pages/docguide/style.md#headings

        $(kubectl -n$NAMESPACE get pod -oname | \
        sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p) \
        -- bash -c "rm -rf /opt/sonarqube/data/es5/* "
```

This shell script will delete SonarQube pod and it will be rescheduled
```shell
kubectl --namespace $NAMESPACE delete pod $(kubectl -n$NAMESPACE get pod -oname | \
        sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p)
```

# Uninstall the Application

## Using the Google Cloud Platform Console

1. In the GCP Console, open [Kubernetes Applications](https://console.cloud.google.com/kubernetes/application).

1. From the list of applications, click **Sonarqube-1**.

1. On the Application Details page, click **Delete**.

## Using the command line

### Prepare the environment

Set your installation name and Kubernetes namespace:

```shell
export APP_INSTANCE_NAME=sonarqube-1
export NAMESPACE=default
```

### Delete the resources

> **NOTE:** We recommend to use a kubectl version that is the same as the version of your cluster. Using the same versions of kubectl and the cluster helps avoid unforeseen issues.

To delete the resources, use the expanded manifest file used for the
installation.

> 

Reviewer 2: ```suggestion
```

Run `kubectl` on the expanded manifest file:

```shell
kubectl delete -f ${APP_INSTANCE_NAME}_manifest.yaml --namespace $NAMESPACE
```

Otherwise, delete the resources using types and a label:

```shell

> kubectl delete application,statefulset,service \

Reviewer 2: This application uses more resources than just those 3 listed. Please add them.

  --namespace $NAMESPACE \
  --selector app.kubernetes.io/name=$APP_INSTANCE_NAME
```

### Delete the persistent volumes of your installation

By design, the removal of StatefulSets in Kubernetes does not remove
PersistentVolumeClaims that were attached to their Pods. This prevents your
installations from accidentally deleting stateful data.

To remove the PersistentVolumeClaims with their attached persistent disks, run
the following `kubectl` commands:

```shell
for pv in $(kubectl get pvc --namespace $NAMESPACE \
  --selector app.kubernetes.io/name=$APP_INSTANCE_NAME \
  --output jsonpath='{.items[*].spec.volumeName}');
do
  kubectl delete pv/$pv --namespace $NAMESPACE
done

kubectl delete persistentvolumeclaims \
  --namespace $NAMESPACE \
  --selector app.kubernetes.io/name=$APP_INSTANCE_NAME
```

### Delete the GKE cluster

Optionally, if you don't need the deployed application or the GKE cluster,
delete the cluster using this command:

```
gcloud container clusters delete "$CLUSTER" --zone "$ZONE"

> ```

Reviewer 2: rm blank line

> 

Reviewer 2: rm blank line

```shell
kubectl --namespace $NAMESPACE cp $(kubectl -n$NAMESPACE get pod -oname | \
                sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p):/opt/sonarqube/data ./
```

## Backing up PostgreSQL

This shell script will create `postgresql-backup.sql` dump of all DB in PostgreSQL.

```shell
kubectl --namespace $NAMESPACE exec -t \
	$(kubectl -n$NAMESPACE get pod -oname | \
		sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
	-c postgresql-server \
	-- pg_dumpall -c -U postgres > backup.sql
```

## Restoring your PostgreSQL

This shell script will prepare dump for restore

```shell
sed -i '1 i\REVOKE CONNECT ON DATABASE sonar FROM PUBLIC;' backup.sql
sed -i '1 a select pg_terminate_backend(pid) from pg_stat_activity where datname='sonar';' backup.sql
echo "GRANT CONNECT ON DATABASE sonar TO PUBLIC;" >> backup.sql
```

This shell script will restore dump `backup.sql` to PostgreSQL

```shell
cat backup.sql | kubectl --namespace $NAMESPACE exec -i \
	$(kubectl -n$NAMESPACE get pod -oname | \
		sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
	-c postgresql-server \
	-- psql -U postgres
```

> This shell script will copy files from current folder to folder `data` in pod

Reviewer: Here we are string restoirng SonarQube status, probably this should have some SonarQube specific title?

Author: part "Restoring" was rewritten

> This shell script will copy files from current folder to folder `data` in pod

Reviewer: I do not like `pod sonarqube`. `sonarqube` is not a name of that pod, it's a pod running SonarQube application.

Author: part "Restoring" was rewritten

> sonarqube

Reviewer: Maybe it's worth packing all these steps into `sonarqube-restore.sh` script? No description for these steps will be needed, as everything will be put into script. Please check [k8s/cassandra/scripts](https://github.com/GoogleCloudPlatform/click-to-deploy/blob/master/k8s/cassandra/scripts/restore.sh) for sample.

If you prefer to leave this in Readme, please provide more verbose description what is happening.

Author: part "Restoring" was rewritten

An Application resource is a collection of individual Kubernetes components,
such as Services, Deployments, and so on, that you can manage as a group.

To set up your cluster to understand Application resources, run the following command:

```shell
kubectl apply -f "https://raw.githubusercontent.com/GoogleCloudPlatform/marketplace-k8s-app-tools/master/crd/app-crd.yaml"
```

You need to run this command once for each cluster.

The Application resource is defined by the
[Kubernetes SIG-apps](https://github.com/kubernetes/community/tree/master/sig-apps) community. The source code can be found on
[github.com/kubernetes-sigs/application](https://github.com/kubernetes-sigs/application).

### Install the application

Navigate to the `sonarqube` directory:

```shell
cd click-to-deploy/k8s/sonarqube
```

#### Configure the application with environment variables

Choose an instance name and
[namespace](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/)
for the application. In most cases, you can use the `default` namespace.

```shell
export APP_INSTANCE_NAME=sonarqube-1
export NAMESPACE=default
```

Configure the container image:

```shell

> export IMAGE_SONARQUBE="marketplace.gcr.io/google/sonarqube7:7.6"

Reviewer 2: Please use republished images instead.

For example:

```shell
TAG=7.6
export IMAGE_SONARQUBE="marketplace.gcr.io/google/sonarqube:${TAG}"
export IMAGE_POSTGRESQL="marketplace.gcr.io/google/sonarqube/postgresql9:${TAG}"
```

And so on.

Author: Updated

  --set postgresql.image=$IMAGE_POSTGRESQL \
  --set postgresql.exporter.image=$IMAGE_POSTGRESQL_EXPORTER \
  --set postgresql.db.password=$POSTGRESQL_DB_PASSWORD  \
  --set metrics.image=$METRICS_EXPORTER_ENABLED > ${APP_INSTANCE_NAME}_manifest.yaml
```
#### Apply the manifest to your Kubernetes cluster

Use `kubectl` to apply the manifest to your Kubernetes cluster:

```shell
kubectl apply -f "${APP_INSTANCE_NAME}_manifest.yaml" --namespace "${NAMESPACE}"
```

#### View the app in the Google Cloud Console

To get the GCP Console URL for your app, run the following command:

```shell
echo "https://console.cloud.google.com/kubernetes/application/${ZONE}/${CLUSTER}/${NAMESPACE}/${APP_INSTANCE_NAME}"
```

To view the app, open the URL in your browser.

# Using the SonarQube Community Edition

By default, the application is not exposed externally. To get access to SonarQube, run the following command:

```bash
kubectl port-forward --namespace $NAMESPACE svc/$APP_INSTANCE_NAME-sonarqube-svc 9000:9000
```

Then, navigate to the [http://localhost:9000/metrics](http://localhost:9000/metrics) endpoint. Use the username `admin` and password `admin` to login.

> To get access web-page with default credentials:

Reviewer 2: Can you delete L224-229, as we already included credentials in the sentence above?

Author: Done

All data and extensions of SonarQube and PostgreSQL stored on PVC which makes the application more stable.

# Installation

## Quick install with Google Cloud Marketplace

Get up and running with a few clicks! Install this Sample Application to a Google Kubernetes Engine cluster using Google Cloud Marketplace. Follow the
[on-screen instructions](https://console.cloud.google.com/marketplace/details/google/sonarqube).

## Command line instructions

You can use [Google Cloud Shell](https://cloud.google.com/shell/) or a local workstation in the
further instructions.

[![Open in Cloud Shell](http://gstatic.com/cloudssh/images/open-btn.svg)](https://console.cloud.google.com/cloudshell/editor?cloudshell_git_repo=https://github.com/GoogleCloudPlatform/click-to-deploy&cloudshell_working_dir=k8s/sonarqube)

### Prerequisites

#### Set up command line tools

You'll need the following tools in your environment:

- [gcloud](https://cloud.google.com/sdk/gcloud/)
- [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)
- [docker](https://docs.docker.com/install/)

> - [openssl](https://www.openssl.org/)

Author: added

such as Services, Deployments, and so on, that you can manage as a group.

To set up your cluster to understand Application resources, run the following command:

```shell
kubectl apply -f "https://raw.githubusercontent.com/GoogleCloudPlatform/marketplace-k8s-app-tools/master/crd/app-crd.yaml"
```

You need to run this command once for each cluster.

The Application resource is defined by the
[Kubernetes SIG-apps](https://github.com/kubernetes/community/tree/master/sig-apps) community. The source code can be found on
[github.com/kubernetes-sigs/application](https://github.com/kubernetes-sigs/application).

### Install the application

Navigate to the `sonarqube` directory:

```shell
cd click-to-deploy/k8s/sonarqube
```

#### Configure the application with environment variables

Choose an instance name and
[namespace](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/)
for the application. In most cases, you can use the `default` namespace.

```shell
export APP_INSTANCE_NAME=sonarqube-1
export NAMESPACE=default
```

Configure the container image:

```shell
TAG=7.6
export IMAGE_SONARQUBE="marketplace.gcr.io/google/sonarqube:$TAG"

> export IMAGE_POSTGRESQL="marketplace.gcr.io/google/sonarqube/postgresql9:$TAG"

Author: fixed

export IMAGE_POSTGRESQL_EXPORTER="marketplace.gcr.io/google/sonarqube/postgresql9-exporter:$TAG"
export IMAGE_METRICS_EXPORTER="marketplace.gcr.io/google/sonarqube/prometheus-to-sd:$TAG"
```

The image above is referenced by
[tag](https://docs.docker.com/engine/reference/commandline/tag). We recommend
that you pin each image to an immutable
[content digest](https://docs.docker.com/registry/spec/api/#content-digests).
This ensures that the installed application always uses the same images,
until you are ready to upgrade. To get the digest for the image, use the
following script:

```shell
for i in "IMAGE_SONARQUBE" "IMAGE_POSTGRESQL" "IMAGE_POSTGRESQL_EXPORTER" "IMAGE_METRICS_EXPORTER"; do
  repo=$(echo ${!i} | cut -d: -f1);
  digest=$(docker pull ${!i} | sed -n -e 's/Digest: //p');
  export $i="$repo@$digest";
  env | grep $i;
done
```

Create a certificate for PostgreSQL. If you already have a certificate that you

> want to use, copy your certificate and key pair in to the `server.crt` and

Author: fixed

openssl req -x509 -nodes -days 365 -newkey rsa:2048 \
    -keyout /tmp/server.key \
    -out /tmp/server.crt \
    -subj "/CN=postgresql/O=postgresql"

kubectl --namespace $NAMESPACE create secret generic $APP_INSTANCE_NAME-tls \
        --from-file=/tmp/server.crt --from-file=/tmp/server.key
```

Generate random password for PosgreSQL:

```shell
export POSTGRESQL_DB_PASSWORD=$(openssl rand 9 | openssl base64 -A | openssl base64)
```

Enable Stackdriver Metrics Exporter:

> **NOTE:** Your GCP project should have Stackdriver enabled. For non-GCP clusters, export of metrics to Stackdriver is not supported yet.

By default the integration is disabled. To enable, change the value to `true`.

```shell
export METRICS_EXPORTER_ENABLED=false
```

#### Expand the manifest template

Use `helm template` to expand the template. We recommend that you save the
expanded manifest file for future updates to the application.

```shell
helm template chart/sonarqube \
  --name="$APP_INSTANCE_NAME" \
  --namespace="$NAMESPACE" \
  --set sonarqube.image=$IMAGE_SONARQUBE \
  --set postgresql.image=$IMAGE_POSTGRESQL \
  --set postgresql.exporter.image=$IMAGE_POSTGRESQL_EXPORTER \

>   --set postgresql.db.password=$POSTGRESQL_DB_PASSWORD  \

Author: fixed

  --set metrics.image=$METRICS_EXPORTER_ENABLED > ${APP_INSTANCE_NAME}_manifest.yaml
```
#### Apply the manifest to your Kubernetes cluster

Use `kubectl` to apply the manifest to your Kubernetes cluster:

```shell
kubectl apply -f "${APP_INSTANCE_NAME}_manifest.yaml" --namespace "${NAMESPACE}"
```

#### View the app in the Google Cloud Console

To get the GCP Console URL for your app, run the following command:

```shell
echo "https://console.cloud.google.com/kubernetes/application/${ZONE}/${CLUSTER}/${NAMESPACE}/${APP_INSTANCE_NAME}"
```

To view the app, open the URL in your browser.

# Using the SonarQube Community Edition

By default, the application is not exposed externally. To get access to SonarQube, run the following command:

```bash
kubectl port-forward --namespace $NAMESPACE svc/$APP_INSTANCE_NAME-sonarqube-svc 9000:9000
```

Then, navigate to the [http://localhost:9000/metrics](http://localhost:9000/metrics) endpoint. Use the username `admin` and password `admin` to login.

> 

Author: this information mentioned on web page of application it is not necessary to duplicate it.

# Application metrics

## Prometheus metrics

The application is configured to expose its metrics through
[PostgreSQL Server Exporter](https://github.com/wrouesnel/postgres_exporter) in
the
[Prometheus format](https://github.com/prometheus/docs/blob/master/content/docs/instrumenting/exposition_formats.md).

You can access the metrics at `[POSTGRESQL_CLUSTER_IP]:9187/metrics`, where
`[POSTGRESQL_CLUSTER_IP]` is the IP address of the application on Kubernetes
cluster.

> 

Author: added

metrics are automatically exported to Stackdriver and visible in
[Stackdriver Metrics Explorer](https://cloud.google.com/monitoring/charts/metrics-explorer).
The name of each metric starts with the application's name, which you define in
the `APP_INSTANCE_NAME` environment variable.

The exporting option might not be available for GKE on-prem clusters.

> Note: Stackdriver has [quotas](https://cloud.google.com/monitoring/quotas) for
> the number of custom metrics created in a single GCP project. If the quota is
> met, additional metrics might not show up in the Stackdriver Metrics Explorer.

You can remove existing metric descriptors using
[Stackdriver's REST API](https://cloud.google.com/monitoring/api/ref_v3/rest/v3/projects.metricDescriptors/delete).

# Scaling

SonarQube Community Edition doest not support scaling.

# Backup and restore

There are 4 core components that are building SonarQube Platform:
- SonarQube Server responsible for starting 3 process which include:
  - A web server allowing to configure SonarQube instance
  - A search server which is based on Elasticsearch
  - A compute engine server for analysis and processing
- SonarQube Database
- A set of plugins

> - SonarScanners which are responsible for project analysis - those usually run on build servers

Author: fixed

For our application database the most important place, `plugins` folder stored on PVC (Persistent Volume Claim).

## Backing up plugin and data

For backup pluging you have to copy files from folder `$SONARQUBE_HOME/extension`.
This shell script will create copy of plugins folder in current folder

```shell
kubectl --namespace $NAMESPACE cp $(kubectl -n$NAMESPACE get pod -oname | \
                sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p):/opt/sonarqube/extensions/ ./
```

This shell script will create copy of content from folder `data` in current folder:

```shell
kubectl --namespace $NAMESPACE cp $(kubectl -n$NAMESPACE get pod -oname | \
                sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p):/opt/sonarqube/data ./
```

> it is not necessary to backup data folder it will be downlnoaded by application

Author: It means if you delete folder data it will be generated again after restart

> it is not necessary to backup data folder it will be downlnoaded by application

Author: fixed

from database.

## Backing up PostgreSQL

> All configuration and data about projects stored in database.

Author: fixed

This shell script will create `postgresql-backup.sql` dump of all DB in PostgreSQL.

```shell
kubectl --namespace $NAMESPACE exec -t \
	$(kubectl -n$NAMESPACE get pod -oname | \
		sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
	-c postgresql-server \

> 	-- pg_dumpall -c -U postgres > backup.sql

Author: as it was explained  `backup.sql` above after fix

```

## Restoring

> Entering to maintenance mode:

Author: It means kind of maintenance mode fore restore because you have to `lock` schema to restore it.

In order to restore PostgresSQL database there is a need to perform some preliminary steps:

> The command below will “lock” the database by blocking new incoming connections:

Author: fixed

```shell
kubectl --namespace $NAMESPACE exec -t \
  $(kubectl -n$NAMESPACE get pod -oname | \
     sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
  -c postgresql-server \
  -- psql -U postgres -c "REVOKE CONNECT ON DATABASE sonar FROM PUBLIC;"
```

Next in order to ensure data consistency all active connections will be dropped:

```shell
kubectl --namespace $NAMESPACE exec -t \
  $(kubectl -n$NAMESPACE get pod -oname | \
     sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
  -c postgresql-server \
  -- psql -U postgres -c "select pg_terminate_backend(pid) from pg_stat_activity where datname='sonar';"
```

Now you are able to restore data to the database.

> Below shell script will restore data from `backup.sql` to PostgreSQL

Author: fixed

```shell
cat backup.sql | kubectl --namespace $NAMESPACE exec -i \
  $(kubectl -n$NAMESPACE get pod -oname | \
    sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
  -c postgresql-server \
  -- psql -U postgres
```

> Next copy files from current folder to folder `data` in SonarQube application pod:

Reviewer 2: Instead of writing next, afterwards prefixes. Could you use a list?

For example:

1.  Run the command below to **lock** the database by blocking new incoming connections:

    ```shell
    kubectl --namespace $NAMESPACE exec -t \
      $(kubectl -n$NAMESPACE get pod -oname | \
      sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
       -c postgresql-server \
          -- psql -U postgres -c "REVOKE CONNECT ON DATABASE sonar FROM PUBLIC;"
    ```

1.  Ensure data consistency, all active connections will be dropped:

    ```shell
    kubectl --namespace $NAMESPACE exec -t \
      $(kubectl -n$NAMESPACE get pod -oname | \
      sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
       -c postgresql-server \
          -- psql -U postgres -c "select pg_terminate_backend(pid) from pg_stat_activity where datname='sonar';"
    ```

1.  Restore data to the database. Below command restores data from backup.sql to PostgreSQL:

    ```shell
    cat backup.sql | kubectl --namespace $NAMESPACE exec -i \
      $(kubectl -n$NAMESPACE get pod -oname | \
      sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
       -c postgresql-server \
          -- psql -U postgres
    ```

Author: fixed

```shell
kubectl --namespace $NAMESPACE cp ./ $(kubectl -n$NAMESPACE get pod -oname | \
  sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p):/opt/sonarqube/data
```

> Afterwards copy files from current folder to `extensions` folder in SonarQube application pod

Author: fixed

```shell
kubectl --namespace $NAMESPACE cp ./ $(kubectl -n$NAMESPACE get pod -oname | \
  sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p):/opt/sonarqube/extensions
```

Delete all outstanding and unneeded SonarQube application data:

```shell

> kubectl --namespace $NAMESPACE exec -i \

Reviewer: This command failed.

Author: it works fine

  $(kubectl -n$NAMESPACE get pod -oname | \
    sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p) \
  -- bash -c "rm -rf /opt/sonarqube/data/es5/* "
```

Exit maintenance mode by unlocking schema `sonar` for new incoming connection:

```shell
kubectl --namespace $NAMESPACE exec -t \
  $(kubectl -n$NAMESPACE get pod -oname | \
    sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
  -c postgresql-server \
  -- psql -U postgres -c "GRANT CONNECT ON DATABASE sonar TO PUBLIC;"
```

Finally restart SonarQube application pod:

```shell
kubectl --namespace $NAMESPACE delete pod $(kubectl -n$NAMESPACE get pod -oname | \
  sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p)
```

> 

Reviewer: After following this procedure, I was unable to connect to SonarQube, with following error on port-forward
```
E0429 16:02:32.162420  260600 portforward.go:331] an error occurred forwarding 9000 -> 9000: error forwarding port 9000 to pod 616d364008da583808ae0a8722899210409ea018f8d3ff2de55ce5c45d4ec879, uid : exit status 1: 2019/04/29 14:02:32 socat[3797717] E connect(5, AF=2 127.0.0.1:9000, 16): Connection refused
```

Author: it is normal behavior you dint need port-forwarding for backup-restore procedure

# Uninstall the Application

## Using the Google Cloud Platform Console

1. In the GCP Console, open [Kubernetes Applications](https://console.cloud.google.com/kubernetes/application).

1. From the list of applications, click **Sonarqube-1**.

1. On the Application Details page, click **Delete**.

## Using the command line

### Prepare the environment

Set your installation name and Kubernetes namespace:

```shell
export APP_INSTANCE_NAME=sonarqube-1
export NAMESPACE=default
```

### Delete the resources

> **NOTE:** We recommend to use a kubectl version that is the same as the version of your cluster. Using the same versions of kubectl and the cluster helps avoid unforeseen issues.

To delete the resources, use the expanded manifest file used for the
installation.

Run `kubectl` on the expanded manifest file:

```shell
kubectl delete -f ${APP_INSTANCE_NAME}_manifest.yaml --namespace $NAMESPACE
```

Otherwise, delete the resources using types and a label:

```shell

> kubectl delete application,deployment,service,pvc \

Reviewer: How about removing secret? Especially tls secret?

--set postgresql.image=$IMAGE_POSTGRESQL \
--set postgresql.exporter.image=$IMAGE_POSTGRESQL_EXPORTER \
--set postgresql.db.password=$POSTGRESQL_DB_PASSWORD \
--set metrics.image=$METRICS_EXPORTER_ENABLED > ${APP_INSTANCE_NAME}_manifest.yaml
```
#### Apply the manifest to your Kubernetes cluster

Use `kubectl` to apply the manifest to your Kubernetes cluster:

```shell
kubectl apply -f "${APP_INSTANCE_NAME}_manifest.yaml" --namespace "${NAMESPACE}"
```

#### View the app in the Google Cloud Console

To get the GCP Console URL for your app, run the following command:

```shell
echo "https://console.cloud.google.com/kubernetes/application/${ZONE}/${CLUSTER}/${NAMESPACE}/${APP_INSTANCE_NAME}"
```

To view the app, open the URL in your browser.

# Using the SonarQube Community Edition

By default, the application is not exposed externally. To get access to SonarQube, run the following command:

```bash
kubectl port-forward --namespace $NAMESPACE svc/$APP_INSTANCE_NAME-sonarqube-svc 9000:9000
```

Then, navigate to the [http://localhost:9000/](http://localhost:9000/) endpoint. Use the username `admin` and password `admin` to login.

# Application metrics

> ## Prometheus metrics

Reviewer 2: Following the header below, this probably should be "PostgreSQL metrics".

Author: Updated

    ```

1. Next in order to ensure data consistency all active connections will be dropped:

    ```shell
    kubectl --namespace $NAMESPACE exec -t \
      $(kubectl -n$NAMESPACE get pod -oname | \
         sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
      -c postgresql-server \
      -- psql -U postgres -c "select pg_terminate_backend(pid) from pg_stat_activity where datname='sonar';"
    ```

1. Now you are able to restore data to the database.
Below shell script will restore data from `postgresql/backup.sql` to PostgreSQL

    ```shell
    cat postgresql/backup.sql | kubectl --namespace $NAMESPACE exec -i \
      $(kubectl -n$NAMESPACE get pod -oname | \
        sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
      -c postgresql-server \
      -- psql -U postgres
    ```

1. Next copy files from current folder to folder `$SONARQUBE_HOME/data` in SonarQube application pod:

    ```shell
    kubectl --namespace $NAMESPACE cp data $(kubectl -n$NAMESPACE get pod -oname | \
      sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p):/opt/sonarqube/data
    ```

> > **NOTE:** It is optional to restore data folder.

Reviewer 2: ```suggestion
    > **NOTE:** It is optional to restore data folder.
```

1. Afterwards copy files from current folder to `$SONARQUBE_HOME/extensions` folder in SonarQube application pod

    ```shell
    kubectl --namespace $NAMESPACE cp extensions $(kubectl -n$NAMESPACE get pod -oname | \
      sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p):/opt/sonarqube/extensions
    ```

1. Delete all outstanding and unneeded SonarQube application data:

    ```shell
    kubectl --namespace $NAMESPACE exec -i \
      $(kubectl -n$NAMESPACE get pod -oname | \
        sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p) \
      -- bash -c "rm -rf /opt/sonarqube/data/es5/* "
    ```

1. Exit maintenance mode by unlocking schema `sonar` for new incoming connection:

    ```shell
    kubectl --namespace $NAMESPACE exec -t \
      $(kubectl -n$NAMESPACE get pod -oname | \
        sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
      -c postgresql-server \
      -- psql -U postgres -c "update pg_database set datallowconn = true where datname = 'sonar';"
    ```

1. Patch secret to restore password:

   ```shell
   kubectl --namespace $NAMESPACE patch secret sonarqube-1-secret -p '{"data": {"password": "'"$ENCODED_PASS"'"}}'
   ```

>    where `$ENCODED_PASS` value for secret.

Reviewer 2: ```suggestion
   where `$ENCODED_PASS` is variable with you password.
```

> Note: Stackdriver has [quotas](https://cloud.google.com/monitoring/quotas) for
> the number of custom metrics created in a single GCP project. If the quota is
> met, additional metrics might not show up in the Stackdriver Metrics Explorer.

You can remove existing metric descriptors using
[Stackdriver's REST API](https://cloud.google.com/monitoring/api/ref_v3/rest/v3/projects.metricDescriptors/delete).

# Scaling

SonarQube Community Edition doest not support scaling.

# Backup and restore

There are 4 core components that are building SonarQube Platform:
- SonarQube Server responsible for starting 3 process which include:
- A web server allowing to configure SonarQube instance
- A search server which is based on Elasticsearch
- A compute engine server for analysis and processing
- SonarQube Database
- A set of plugins
- SonarQube Scanners which are responsible for project analysis - those usually run on build servers
For our application database the most important place, `plugins` folder stored on PVC (Persistent Volume Claim).

## Backing up plugin and data

For backup pluging you have to copy files from folder `$SONARQUBE_HOME/extension`.

> This shell script will create copy of plugins folder into `ext` folder:

Author: fixed

```shell
mkdir ext
cd ext
kubectl --namespace $NAMESPACE cp $(kubectl -n$NAMESPACE get pod -oname | \
              sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p):/opt/sonarqube/extensions/ ./
cd ..

> ```

Author: fixed

This shell script will create copy of content from folder `$SONARQUBE_HOME/data` into folder `data`:

```shell
mkdir data
cd data
kubectl --namespace $NAMESPACE cp $(kubectl -n$NAMESPACE get pod -oname | \

>               sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p):/opt/sonarqube/data ./

Author: fixed

cd ..
```

It is not necessary to backup data folder, it will be downlnoaded by application
from database.

## Backing up PostgreSQL

All configuration and data about projects are stored in database.

> This shell script will create `backup.sql` dump of all DB in PostgreSQL.

Author: fixed

> ```shell

Author: fixed

mkdir postgresql
cd postgresql
kubectl --namespace $NAMESPACE exec -t \
	$(kubectl -n$NAMESPACE get pod -oname | \
		sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
	-c postgresql-server \
	-- pg_dumpall -c -U postgres > backup.sql
cd ..
```

## Backup Password

> This shell script will show you encrypted password to PostgreSQL

Author: yep my fail, fixed

```shell
kubectl get secret $APP_INSTANCE_NAME-secret --namespace $NAMESPACE -o yaml | grep password:
```

> **NOTE:** It is important to backup password.

## Restoring

Entering to maintenance mode:
In order to restore PostgresSQL database there is a need to perform some preliminary steps:

1. The command below will “lock” the database by blocking new incoming
    connections:

    ```shell
    kubectl --namespace $NAMESPACE exec -t \
      $(kubectl -n$NAMESPACE get pod -oname | \
         sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
      -c postgresql-server \
      -- psql -U postgres -c "update pg_database set datallowconn = false where datname = 'sonar';"
    ```

1. Next in order to ensure data consistency all active connections will be dropped:

    ```shell
    kubectl --namespace $NAMESPACE exec -t \
      $(kubectl -n$NAMESPACE get pod -oname | \
         sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
      -c postgresql-server \
      -- psql -U postgres -c "select pg_terminate_backend(pid) from pg_stat_activity where datname='sonar';"
    ```

1. Now you are able to restore data to the database.

> Below shell script will restore data from `backup.sql` to PostgreSQL

Author: fixed

    ```shell

>     cat backup.sql | kubectl --namespace $NAMESPACE exec -i \

Author: fixed

      $(kubectl -n$NAMESPACE get pod -oname | \
        sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
      -c postgresql-server \
      -- psql -U postgres
    ```

> 1. Next copy files from current folder to folder `$SONARQUBE_HOME/data` in SonarQube application pod:

Author: Yes it is optional I will add some notice about this

    ```shell

>     kubectl --namespace $NAMESPACE cp ./ $(kubectl -n$NAMESPACE get pod -oname | \

Author: fixed

      sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p):/opt/sonarqube/data
    ```

1. Afterwards copy files from current folder to `$SONARQUBE_HOME/extensions` folder in SonarQube application pod

>     ```shell

Author: fixed

    kubectl --namespace $NAMESPACE cp ./ $(kubectl -n$NAMESPACE get pod -oname | \
      sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p):/opt/sonarqube/extensions
    ```

1. Delete all outstanding and unneeded SonarQube application data:

    ```shell
    kubectl --namespace $NAMESPACE exec -i \
      $(kubectl -n$NAMESPACE get pod -oname | \
        sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p) \
      -- bash -c "rm -rf /opt/sonarqube/data/es5/* "
    ```

1. Exit maintenance mode by unlocking schema `sonar` for new incoming connection:

    ```shell
    kubectl --namespace $NAMESPACE exec -t \
      $(kubectl -n$NAMESPACE get pod -oname | \
        sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
      -c postgresql-server \
      -- psql -U postgres -c "update pg_database set datallowconn = true where datname = 'sonar';"
    ```

1. Patch secret to restore password:

   ```shell

>    kubectl patch secret sonarqube-1-secret -p '{"data": {"password": "$ENCRYPTED_PASS"}}'

Reviewer: Please add `--namespace $NAMESPACE`

Author: fixed

   ```

>    where `$ENCRYPTED_PASS` value for secret.

Reviewer: `$ENCRYPTED_PASS` is base64 encoded password, there is no encryption happening. Please change variable name.

Also, using `'` prevents `"$ENCRYPTED_PASS"` from evaluating - please provide a note about this, or change a line to `'{"data": {"password": "'"$ENCRYPTED_PASS"'"}}'`

Author: line changed

Configure the container image:

```shell
TAG=7.6
export IMAGE_SONARQUBE="marketplace.gcr.io/google/sonarqube:$(TAG)"
export IMAGE_POSTGRESQL="marketplace.gcr.io/google/sonarqube/postgresql9:$(TAG)"
export IMAGE_POSTGRESQL_EXPORTER="marketplace.gcr.io/google/sonarqube/postgresql9/exporter:$(TAG)"

> export IMAGE_METRICS_EXPORTER="k8s.gcr.io/prometheus-to-sd:v0.5.1"

Reviewer 2: Is this image republished? If so, could you also update it?

https://github.com/GoogleCloudPlatform/click-to-deploy/blob/af55f617c13cea3463188671c575b1a17952d570/k8s/sonarqube/Makefile#L92-L99

Author: Updated
